File: Leetcode/876/876.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        auto ptr = head;
        int length = 0;
        while (ptr) {
            ++length;
            ptr = ptr->next;
        }
        int half = length / 2 + 1;
        ptr = head;
        while (--half) {
            ptr = ptr -> next;
        }
        return ptr;
    }
};

File: Leetcode/876/876.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNode(head *ListNode) *ListNode {
	node := head
	counter := 0
	for node != nil {
		counter++
		node = node.Next
	}
	half := counter >> 1
	var prev = new(ListNode)
	prev.Next = head
    fmt.Println(half)
	for half > 0 {
		prev = prev.Next
		half--
	}
	return prev.Next
}